src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt

```diff
@@ -18,48 +20,93 @@ import java.net.URI
 
+fun connectToJira(username: String, password: String, url: String) =
+    JiraClient(url, TokenCredentials(username, password))
+
 fun getIssue(jiraClient: JiraClient, key: String) = runBlocking {
     Either.catch {
         jiraClient.getIssue(key, "*all", "changelog")
     }
 }
+
 const val MILLI_FOR_FORMAT = 123L
 
-fun updateCHK(issue: Issue, chkField: String): Either<Throwable, Unit> = runBlocking {
-    Either.catch {
-        issue
-            .transition()
-            .field(
-                chkField,
-                Instant
-                    .now()
-                    .with(ChronoField.NANO_OF_SECOND, 0)
-                    .with(ChronoField.MILLI_OF_SECOND, MILLI_FOR_FORMAT)
-                    .toString()
-                    .replace("Z", "-0000")
-            )
-            .execute("Update Issue")
-    }
+fun updateCHK(context: IssueUpdateContext, chkField: String) {
+    context.hasTransitions = true
+    context.transition.field(
+        chkField,
+        Instant
+            .now()
+            .with(ChronoField.NANO_OF_SECOND, 0)
+            .with(ChronoField.MILLI_OF_SECOND, MILLI_FOR_FORMAT)
+            .toString()
+            .replace("Z", "-0000")
+    )
 }
 
-fun updateConfirmation(issue: Issue, confirmationField: String, value: String) = runBlocking {
-    Either.catch {
-        val jsonValue = JSONObject()
-        jsonValue["value"] = value
+fun updateConfirmation(context: IssueUpdateContext, confirmationField: String, value: String) {
+    val jsonValue = JSONObject()
+    jsonValue["value"] = value
 
-        issue
-            .transition()
-            .field(confirmationField, jsonValue)
-            .execute("Update Issue")
-    }
+    context.hasTransitions = true
+    context.transition.field(confirmationField, jsonValue)
 }
 
-fun updateLinked(issue: Issue, linkedField: String, value: Double) = runBlocking {
-    Either.catch {
-        issue
-            .transition()
-            .field(linkedField, value)
-            .execute("Update Issue")
+fun updateLinked(context: IssueUpdateContext, linkedField: String, value: Double) {
+    context.hasTransitions = true
+    context.transition.field(linkedField, value)
+}
+
+fun reopen(context: IssueUpdateContext) = runBlocking {
+    context.hasTransitions = true
+    context.transitionName = "Reopen Issue"
+}
+
+fun resolveAs(context: IssueUpdateContext, resolution: String) {
+    val resolutionJson = JSONObject()
+    resolutionJson["name"] = resolution
+
+    context.hasTransitions = true
+    context.transition.field(Field.RESOLUTION, resolutionJson)
+    context.transitionName = "Resolve Issue"
+}
+
+fun updateSecurity(context: IssueUpdateContext, levelId: String) {
+    context.hasTransitions = true
+    context.transition.field(Field.SECURITY, Field.valueById(levelId))
+}
+
+fun removeAffectedVersion(context: IssueUpdateContext, version: Version) {
+    context.hasUpdates = true
+    context.update.fieldRemove("versions", version)
```

Comment:
To avoid having both an update transition and a regular update, this could be moved to the update transition instead.
This is true for everything in a regular update, except of the description.

If you'd like to avoid the change being denoted as "Opened", "Reopened" or "Resolved" in emails and logs and and avoid having the status field show up in the changelog, you could do a bit more of an advanced approach and have a registry where fields are regisered with booleans whether it's possible to change them in an update and an update transiton, and prefer an update, when everything is possible as a simple update.

Reply:
:cry: Sadly the Jira library we use doesn't provide a `fieldAdd` and `fieldRemove` method for transition, as I noted in the PR summary

Reply:
This lib also doesn't support deleting attachments, we had to write our own API call for that.
Does the official one not provide the rest client?

Also I think fieldAdd and fieldRemove could still work by setting the field to all previous entries plus (or minus for remove) one entry. Would be kinda annoying to handle for multiple adds/removes, though.

Reply:
No, the official one doesn't even expose the rest client (or at least I can't find it anywhere)

Reply:
> If you'd like to avoid the change being denoted as "Opened", "Reopened" or "Resolved" in emails and logs and and avoid having the status field show up in the changelog

iirc updates don't trigger any email, so we don't even to worry about that, right? And I personally don't really care about that status change shows in the changelog since they are anyways in the same item.

Reply:
> Also I think fieldAdd and fieldRemove could still work by setting the field to all previous entries plus (or minus for remove) one entry.

I gave up lol

Reply:
> 
> 
> > If you'd like to avoid the change being denoted as "Opened", "Reopened" or "Resolved" in emails and logs and and avoid having the status field show up in the changelog
> 
> iirc updates don't trigger any email, so we don't even need to worry about that, right? And I personally don't really care about that status change shows in the changelog since they are anyways in the same item.

Yeah, both methods have advantages and disadvantages, that's why I don't have really have a preference.

Reply:
Actually, if we want to add versions in update this, it would probably be the easiest to just use our own fork from the (current) lib and implement these things there, instead of constantly working around them